### PR TITLE
feat(cli): version requirements as arguments

### LINKED ISSUE #38: Stronger typed CLI (parse, don't validate)
suggestion: I'm a fan of [parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).

Can we apply this here, by making the version a `semver::Version` and perhaps by creating a wrapper type for rock names that allow only valid characters?

It would require more involved arg parsing, but it results in nicer errors if users pass in invalid arguments.

Maybe this is for another PR, as it would mean refactoring the `Download` command, too.

_Originally posted by @mrcjkb in https://github.com/nvim-neorocks/rocks/pull/35#discussion_r1751785967_

### LINKED ISSUE #49: [Feature] version constraints for install, download and search commands
Related to #38

### PR DESCRIPTION
**Pull request by mrcjkb**
Closes #38 
Closes #49 

- Uses `LuaPackageReq` as the CLI args.
- Changes `ManifestMetadata` to have `PackageVersion`s as keys, instead of `String`s.

### CODE REVIEW

File: rocks-lib/src/luarock/mod.rs
Comment (mrcjkb) [inline rocks-lib/src/luarock/mod.rs:1]: The `luarock` module seems to have been re-added accidentally during a rebase.

File: rocks-lib/src/manifest/metadata.rs [hunk 1]
```diff
@@ -23,11 +28,12 @@ impl ManifestMetadata {
     pub fn new(manifest: &String) -> Result<Self> {
         let lua = Lua::new();
 
-        lua.load(manifest).exec()?;
+        lua.load(dbg!(manifest)).exec()?;
```
Comment (vhyrro) [inline rocks-lib/src/manifest/metadata.rs:31]: issue: leftover debug call 😅

File: rocks-lib/src/manifest/metadata.rs [hunk 2]
```diff
@@ -42,19 +48,71 @@ impl ManifestMetadata {
...
+            lua_package_req.name().to_owned(),
+            version.to_owned(),
+        ))
+    }
+}
+
+#[derive(serde::Deserialize)]
+pub struct ManifestRockEntry {
+    /// e.g. "linux-x86_64", "rockspec", "src", ...
+    pub arch: String,
```
Comment (vhyrro) [inline rocks-lib/src/manifest/metadata.rs:88]: question: we'd probably benefit from making this an enum too, wouldn't we? Should this be given a TODO for the future? Note that the `Triple` type can be used to represent a "linux-arch" string quite nicely (with serialization and such).

File: rocks-lib/src/lua_package/mod.rs [hunk 3]
```diff
@@ -14,11 +14,14 @@ pub struct LuaPackage {
...
-        })
+    pub fn new(name: PackageName, version: PackageVersion) -> Self {
+        Self { name, version }
+    }
+    pub fn parse(name: String, version: String) -> Result<Self> {
+        Ok(Self::new(
+            PackageName::new(name),
+            PackageVersion::parse(&version)?,
+        ))
     }
```
Comment (vhyrro) [inline rocks-lib/src/lua_package/mod.rs:25]: question: any benefit from doing `impl From<(String, String)>` as a bonus, or is that too much "magic"?

**Review by vhyrro — COMMENTED** [review]
Big fan, makes the API so much more robust now :)

**Comment by vhyrro** [comment]
Oh, there's also the `Rockspec` which still has a `String` as its version object. Should that be addressed in this PR as well?

File: rocks-lib/src/manifest/metadata.rs (hunk 2 above)
Comment (mrcjkb) [reply rocks-lib/src/manifest/metadata.rs:88]: Created #62

File: rocks-lib/src/lua_package/mod.rs (hunk 3 above)
Comment (mrcjkb) [reply rocks-lib/src/lua_package/mod.rs:25]: The implementation would have to panic if the version can't be parsed, as it doesn't return a `Result`.

Comment (mrcjkb) [reply rocks-lib/src/lua_package/mod.rs:26]: We have a `FromStr`, so I'd say yagni to `From` for now.

**Comment by mrcjkb** [comment]
> Oh, there's also the `Rockspec` which still has a `String` as its version object. Should that be addressed in this PR as well?

Done. See the second commit.

**Pull request merged**